packages/gatsby-plugin-manifest/src/gatsby-node.js

```diff
@@ -18,48 +23,64 @@ function generateIcons(icons, srcIcon) {
-    const { icon, ...manifest } = pluginOptions
+exports.onPostBootstrap = async (args, pluginOptions) => {
+  const { icon, ...manifest } = pluginOptions
 
-    // Delete options we won't pass to the manifest.webmanifest.
+  // Delete options we won't pass to the manifest.webmanifest.
 
-    delete manifest.plugins
-    delete manifest.legacy
-    delete manifest.theme_color_in_head
+  delete manifest.plugins
+  delete manifest.legacy
+  delete manifest.theme_color_in_head
+  delete manifest.cache_busting_mode
 
-    // If icons are not manually defined, use the default icon set.
-    if (!manifest.icons) {
-      manifest.icons = defaultIcons
+  // If icons are not manually defined, use the default icon set.
+  if (!manifest.icons) {
+    manifest.icons = defaultIcons
+  }
+
+  // Determine destination path for icons.
+  const iconPath = path.join(`public`, path.dirname(manifest.icons[0].src))
+
+  //create destination directory if it doesn't exist
+  if (!fs.existsSync(iconPath)) {
+    fs.mkdirSync(iconPath)
+  }
+
+  // Only auto-generate icons if a src icon is defined.
+  if (icon !== undefined) {
+    // Check if the icon exists
+    if (!doesIconExist(icon)) {
+      throw `icon (${icon}) does not exist as defined in gatsby-config.js. Make sure the file exists relative to the root of the site.`
     }
 
-    // Determine destination path for icons.
-    const iconPath = path.join(`public`, path.dirname(manifest.icons[0].src))
+    //add cache busting
+    const cacheMode =
+      typeof pluginOptions.cache_busting_mode !== `undefined`
+        ? pluginOptions.cache_busting_mode
+        : `query`
 
-    //create destination directory if it doesn't exist
-    if (!fs.existsSync(iconPath)) {
-      fs.mkdirSync(iconPath)
+    //if cacheBusting is being done via url query icons must be generated before cachebusting runs
+    if (cacheMode === `query`) {
+      await generateIcons(manifest.icons, icon)
     }
 
-    fs.writeFileSync(
-      path.join(`public`, `manifest.webmanifest`),
-      JSON.stringify(manifest)
-    )
-
-    // Only auto-generate icons if a src icon is defined.
-    if (icon !== undefined) {
-      // Check if the icon exists
-      if (!doesIconExist(icon)) {
-        reject(
-          `icon (${icon}) does not exist as defined in gatsby-config.js. Make sure the file exists relative to the root of the site.`
-        )
-      }
-      generateIcons(manifest.icons, icon).then(() => {
-        //images have been generated
-        console.log(`done generating icons for manifest`)
-        resolve()
+    if (cacheMode !== `none`) {
+      let iconDigest = createContentDigest(fs.readFileSync(icon))
+
+      manifest.icons.forEach(icon => {
+        icon.src = addDigestToPath(icon.src, iconDigest, cacheMode)
       })
-    } else {
-      resolve()
     }
-  })
+
+    //if file names are being modified by cacheBusting icons must be generated after cachebusting runs
+    if (cacheMode === `name`) {
```

Comment:
with this and previous `if (cacheMode === `query`)` it seems to me that when using `cacheMode = 'none'` we won't generate icons?

Could we perhaps move generating icon names into `generateIcons` function itself?

Reply:
@pieh Good catch! I'll get this fixed and make sure the tests catch something like this in the future.

Reply:
@pieh I was shying away from that since its not used elsewhere but that doesn't really matter. Tests are more important.

Reply:
Okay, this is a strange issue now. Moving the imports and generate Icon function to `common.js` makes the `gatsby-ssr.js` tests thrown an error.

```
packages/gatsby-plugin-manifest/src/__tests__/gatsby-ssr.js
  ● Test suite failed to run

    TypeError: exists is not a function

      2 | import path from "path"
      3 | import Promise from "bluebird"
    > 4 | import sharp from "sharp"
        | ^
      5 | import crypto from "crypto"
      6 |
      7 | sharp.simd(true)

      at Function.getRoot (packages/gatsby-plugin-manifest/node_modules/bindings/bindings.js:201:7)
      at bindings (packages/gatsby-plugin-manifest/node_modules/bindings/bindings.js:82:32)
      at Object.<anonymous> (packages/gatsby-plugin-manifest/node_modules/sharp/lib/constructor.js:10:34)
      at Object.<anonymous> (packages/gatsby-plugin-manifest/node_modules/sharp/lib/index.js:3:15)
      at Object.<anonymous> (packages/gatsby-plugin-manifest/src/common.js:4:1)
      at Object.<anonymous> (packages/gatsby-plugin-manifest/src/gatsby-ssr.js:5:1)
      at Object.require (packages/gatsby-plugin-manifest/src/__tests__/gatsby-ssr.js:7:26)

```

Reply:
Well I just reverted this modification and targeted `sharp` instead since that's ultimately what's being called by `generateIcons`